[ews-build.webkit.org] Classify commits in pull-requests

JonWBedard opened the pull request:
#### 07098ea5d9313615068f339e7f7471b21bfaa9ba
<pre>
[ews-build.webkit.org] Classify commits in pull-requests
<a href="https://bugs.webkit.org/show_bug.cgi?id=254680">https://bugs.webkit.org/show_bug.cgi?id=254680</a>
rdar://107381451

Reviewed by Aakash Jain.

Many EWS steps preform rudimentry and divergent classification of commits.
For example, merge-queue preforms less validation on gardening commits and reverts.
The WebKit Project has a concept of &apos;commit classification&apos; that `git-webkit` uses
to determine which changes on a given branch are novel. To pave the way for EWS to
use the &apos;commit classification&apos; concept, buildrequests need &apos;classification&apos; as one
of their properties.

* Tools/CISupport/ews-build/events.py:
(logging_disabled.__exit__):
(CommitClassifier): Add a class which encapsulates the concept of a commit classification.
(CommitClassifier.HeaderFilter): Add class encapsulating both a regex and fuzzy match
filter to search the title of commits with.
(GitHubEventHandlerNoEdits.commit_classes): Return a cached list of all CommitClassifier
bjects for the WebKit project.
(GitHubEventHandlerNoEdits._get_commit_msg): Clarify why we&apos;re implementing our own
&quot;_get_commit_messages&quot; and (delibrately) disabling this one.
(GitHubEventHandlerNoEdits._get_commit_messages): Return a list of the commit messages
for the commits in this pull request.
(GitHubEventHandlerNoEdits.classifiy): Given a commit message and list of changed
files, return a string representing the CommitClassifier of the provided commit message,
if the commit message matches one of our commit classes.
(GitHubEventHandlerNoEdits.handle_pull_request): Before returning a change, determine
the classification of all commits in the pull-request, so that information can be
included in properties.
* Tools/CISupport/ews-build/events_unittest.py: Added.
* Tools/CISupport/runUnittests.py:
(main): Explicitly import rapidfuzz, since that is an uncommon dependency.

Canonical link: <a href="https://commits.webkit.org/262332@main">https://commits.webkit.org/262332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1820399d48e8e17e37c1dc5520822218a0244402

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1279 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/1314 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/1354 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/2098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/1146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/1266 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/1367 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/1383 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/2098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/1290 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/1367 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/1354 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/1952 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/1367 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/1354 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/1952 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/1173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/1354 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/1952 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/1221 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/1383 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/28/builds/1269 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/1159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/1354 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/1248 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/126 "Failed to push commit to Webkit repository") | | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -278,6 +280,63 @@ def __exit__(self, type, value, tb):
 
+# Based off of webkitscmpy's CommitClassifier, although modified to not raise exceptions
+# when provided invalid commit classes.
+class CommitClass(object):
```
The name can be more descriptive to indicate what's the purpose of this class, e.g.: CommitClassifier

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -278,6 +280,63 @@ def __exit__(self, type, value, tb):
+
+        if self.headers and not header:
+            return False
+        if self.headers and not any([h(header) for h in self.headers]):
```
can `h(header)` raise exception (is `h` always callable)?  If so, might consider exception handling here.

JonWBedard replied:
`h(header)` should not be able to raise an exception. `self.headers` is a list of `CommitClassifier.HeaderFilter` objects, and while `HeaderFilter` objects can have an undefined `do`, if `do` is undefined, calling a `HeaderFilter` object will return `False`. All exception will be raised when constructing the `CommitClassifier` object in the first place.

aj062 reviewed (approved):
r+ assuming this is well tested

JonWBedard committed:
[ews-build.webkit.org] Classify commits in pull-requests
https://bugs.webkit.org/show_bug.cgi?id=254680
rdar://107381451

Reviewed by Aakash Jain.

Many EWS steps preform rudimentry and divergent classification of commits.
For example, merge-queue preforms less validation on gardening commits and reverts.
The WebKit Project has a concept of 'commit classification' that `git-webkit` uses
to determine which changes on a given branch are novel. To pave the way for EWS to
use the 'commit classification' concept, buildrequests need 'classification' as one
of their properties.

* Tools/CISupport/ews-build/events.py:
(logging_disabled.__exit__):
(CommitClassifier): Add a class which encapsulates the concept of a commit classification.
(CommitClassifier.HeaderFilter): Add class encapsulating both a regex and fuzzy match
filter to search the title of commits with.
(GitHubEventHandlerNoEdits.commit_classes): Return a cached list of all CommitClassifier
bjects for the WebKit project.
(GitHubEventHandlerNoEdits._get_commit_msg): Clarify why we're implementing our own
"_get_commit_messages" and (delibrately) disabling this one.
(GitHubEventHandlerNoEdits._get_commit_messages): Return a list of the commit messages
for the commits in this pull request.
(GitHubEventHandlerNoEdits.classifiy): Given a commit message and list of changed
files, return a string representing the CommitClassifier of the provided commit message,
if the commit message matches one of our commit classes.
(GitHubEventHandlerNoEdits.handle_pull_request): Before returning a change, determine
the classification of all commits in the pull-request, so that information can be
included in properties.
* Tools/CISupport/ews-build/events_unittest.py: Added.
* Tools/CISupport/runUnittests.py:
(main): Explicitly import rapidfuzz, since that is an uncommon dependency.

Canonical link: https://commits.webkit.org/262332@main
